File: async_checker/naive/AsyncChecker.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

public class AsyncChecker {
    static class TaskResult {
        final int latency;
        final int status;
        TaskResult(int latency, int status) {
            this.latency = latency;
            this.status = status;
        }
    }

    private static TaskResult runTask(int id) {
        long seed = (((long)id * 1664525L + 1013904223L) & 0xFFFFFFFFL);
        int latency = (int)(10 + (seed % 990));
        int status = 200;
        if ((seed % 7) == 0) {
            status = ((seed % 3) == 0) ? 429 : 500;
        }
        return new TaskResult(latency, status);
    }

    private static final TaskResult POISON = new TaskResult(-1, -1);

    public static void main(String[] args) throws Exception {
        int tasks = 10000;
        int workers = 16;
        BlockingQueue<TaskResult> resQueue = new ArrayBlockingQueue<>(1000);
        Thread[] threads = new Thread[workers];

        int chunkSize = tasks / workers;
        for (int w = 0; w < workers; w++) {
            final int start = w * chunkSize;
            final int end = (w == workers - 1) ? tasks : (start + chunkSize);
            threads[w] = new Thread(() -> {
                try {
                    for (int i = start; i < end; i++) {
                        resQueue.put(runTask(i));
                    }
                } catch (InterruptedException ignored) {}
            });
            threads[w].start();
        }

        Thread closer = new Thread(() -> {
            try {
                for (Thread t : threads) t.join();
                resQueue.put(POISON);
            } catch (InterruptedException ignored) {}
        });
        closer.start();

        int okCount = 0, rlCount = 0, errCount = 0;
        long totalLat = 0;
        List<Integer> latencies = new ArrayList<>(tasks);

        while (true) {
            TaskResult res = resQueue.take();
            if (res.latency == -1) break;
            latencies.add(res.latency);
            totalLat += res.latency;
            switch (res.status) {
                case 200 -> okCount++;
                case 429 -> rlCount++;
                case 500 -> errCount++;
            }
        }

        closer.join();

        Collections.sort(latencies);
        int p50 = latencies.get((int)(tasks * 0.50));
        int p95 = latencies.get((int)(tasks * 0.95));
        int p99 = latencies.get((int)(tasks * 0.99));

        System.out.printf("Async complete: tasks=%d, ok=%d, rate_limited=%d, errors=%d, latency_sum=%d, p50=%d, p95=%d, p99=%d\n",
                tasks, okCount, rlCount, errCount, totalLat, p50, p95, p99);
    }
}

File: async_checker/naive/async_checker.js
function runTask(id) {
  const seed = Number((BigInt(id) * 1664525n + 1013904223n) & 0xFFFFFFFFn);
  const latency = 10 + (seed % 990);
  let status = 200;
  if ((seed % 7) === 0) {
    status = (seed % 3) === 0 ? 429 : 500;
  }
  return { latency, status };
}

async function main() {
  const tasks = 10000;
  const promises = [];

  for (let i = 0; i < tasks; i++) {
    promises.push(Promise.resolve().then(() => runTask(i)));
  }

  const results = await Promise.all(promises);

  let okCount = 0, rlCount = 0, errCount = 0;
  let totalLat = 0;
  const latencies = new Array(tasks);

  for (let i = 0; i < tasks; i++) {
    const res = results[i];
    latencies[i] = res.latency;
    totalLat += res.latency;
    if (res.status === 200) okCount++;
    else if (res.status === 429) rlCount++;
    else if (res.status === 500) errCount++;
  }

  latencies.sort((a, b) => a - b);
  const p50 = latencies[Math.floor(tasks * 0.50)];
  const p95 = latencies[Math.floor(tasks * 0.95)];
  const p99 = latencies[Math.floor(tasks * 0.99)];

  console.log(`Async complete: tasks=${tasks}, ok=${okCount}, rate_limited=${rlCount}, errors=${errCount}, latency_sum=${totalLat}, p50=${p50}, p95=${p95}, p99=${p99}`);
}

main();
